Added how-to guides (#8624)
Qiskit/qiskit: docs/how_to/create_a_quantum_circuit.rst

########################
Create a quantum circuit
########################

> This guide shows how to initialize a quantum circuit.

Reviewer: ```suggestion
This guide shows you how to create a quantum circuit.
```

There are two ways to create a :class:`~.QuantumCircuit` object:

> * Specifying the number of qubits and bits.
> * Creating :class:`~.QuantumRegister`\ s and :class:`~.ClassicalRegister`\ s.

Reviewer: ```suggestion
1. Specifying the number of qubits and classical bits.
2. Creating :class:`~.QuantumRegister`\ s and :class:`~.ClassicalRegister`\ s and use the registers to initialize a :class:`~.QuantumCircuit`
```

> Create from number of qubits and bits

Reviewer: ```suggestion
Specifying the number of qubits and classical bits
```
I think the subsection title should match or at least use similar words as the two methods mentioned above.

=====================================

> In order to create a :class:`~.QuantumCircuit` by only specifying the number of bits and qubits, you need to follow these steps.

Reviewer: ```suggestion
You can create a :class:`~.QuantumCircuit` by only specifying the number of qubits and classical bits. For example:
```

.. testcode::

    from qiskit import QuantumCircuit

    # Initialize number of qubits and classical bits
    n_qubits = 3
    n_bits = 2

    # Create the circuit
    qc = QuantumCircuit(n_qubits, n_bits)

>     print(qc.draw())

Reviewer: ```suggestion
    qc.draw(output='text')
```

Author: I had to include the `print` because else `doctest` would interpret that as "no output"

.. testoutput::
    :options: +NORMALIZE_WHITESPACE

    q_0: 
        
    q_1: 
        
    q_2: 
        
    c: 2/
     

> If you don't want to include any classical bits, you don't have to write `QuantumCircuit(n_qubits,0)` but you can omit the number of classical bits.

Reviewer: ```suggestion
You can also create a :class:`~.QuantumCircuit` with only qubits and no classical bits, by omitting the number of classical bits:
```

> If you don't want to include any classical bits, you don't have to write `QuantumCircuit(n_qubits,0)` but you can omit the number of classical bits.

Reviewer: ```suggestion
You can also create a :class:`~.QuantumCircuit` with only qubits and no classical bits, by omitting the number of classical bits:
```

.. testcode::

    from qiskit import QuantumCircuit

    qc = QuantumCircuit(n_qubits)

>     print(qc.draw())

Reviewer: ```suggestion
    qc.draw(output='text')
```

.. testoutput::
    :options: +NORMALIZE_WHITESPACE

    q_0: 
        
    q_1: 
        
    q_2: 

Create from quantum and classical registers

> ===========================================

Reviewer: A how to guide should provide tips for users to decide how to choose different options of the same task. Can you comment on when people should use this method and when they should use the simpler method with specifying only number of qubits and classical bits? For example, the simpler method can only create circuits with one quantum register and one classical register. And it also can't specify names to the registers.

Create quantum registers
------------------------

> In order to create a quantum register, you have to define a :class:`~.QuantumRegister` object, passing as argument the desired number of qubits.

Reviewer: ```suggestion
You can create a :class:`~.QuantumRegister` object by passing the desired number of qubits as an argument:
```

.. testcode::

    from qiskit import QuantumRegister

>     # Create QuantumRegister formed by 2 qubits
>     qr1 = QuantumRegister(2)
> 
>     # Create QuantumRegister formed by 3 qubits
>     qr2 = QuantumRegister(3)

Reviewer: ```suggestion
    # Create a quantum register with 2 qubits
    qr1 = QuantumRegister(2)

    # Create a quantum register with 3 qubits
    qr2 = QuantumRegister(3)
```

Create classical registers
--------------------------

> Analogously to the quantum registers, a classical register can be created by defining a :class:`~.ClassicalRegister` object, passing the number of classical bits as an argument.

Reviewer: ```suggestion
Similar to the quantum registers, you can create a :class:`~.ClassicalRegister` object by passing the desired number of classical bits as an argument:
```

.. testcode::

    from qiskit import ClassicalRegister

>     # Create classical register with 2 classical bits
>     cr1 = ClassicalRegister(2)
> 
>     # Create classical register with 1 classical bit
>     cr2 = ClassicalRegister(1)

Reviewer: ```suggestion
    # Create a classical register with 2 classical bits
    cr1 = ClassicalRegister(2)

    # Create a classical register with 1 classical bit
    cr2 = ClassicalRegister(1)
```

> Initialize the quantum circuit

Reviewer 2: I think it would be good to align the language used in this headings, so either:

> Creating quantum registers
...
> Creating classical registers
...
> Create a quantum circuit
...

or:

> Initializing quantum registers
...
> Initializing classical registers
...
> Initialize a quantum circuit
...

------------------------------

> Now that you have defined the quantum and classical registers, you can define a :class:`~.QuantumCircuit` from them. Each register has to be introduced as a separate argument.

Reviewer: ```suggestion
Now that you have defined the quantum and classical registers, you can create a :class:`~.QuantumCircuit` with the registers: 
```

.. testcode::

>     # Create the quantum circuit from the registers
>     qc = QuantumCircuit(qr1, qr2, cr1, cr2)
>     print(qc.draw())

Reviewer: ```suggestion
    # Create a quantum circuit with the registers
    qc = QuantumCircuit(qr1, qr2, cr1, cr2)
    qc.draw(output='text')
```

.. testoutput::
    :options: +NORMALIZE_WHITESPACE

    q0_0: 
      
    q0_1: 
        
    q1_0: 
        
    q1_1: 
        
    q1_2: 
        
    c0: 2/
        
    c1: 1/
      

> You can put the registers in any order, even mixing classical and quantum. However, the relative order of the :class:`~.QuantumRegister`\ s does affect the order of the qubits on the final circuit. In particular, the qubits from the first :class:`~.QuantumRegister` will be the first and so on. The same applies to the :class:`~.ClassicalRegister`\ s.

Reviewer: ```suggestion
You can put the registers in any order, even mixing classical and quantum. However, the relative order of the :class:`~.QuantumRegister`\ s affects the order of the qubits in the final circuit. The qubits from the first :class:`~.QuantumRegister` will be the first and so on. The same applies to the :class:`~.ClassicalRegister`\ s.
```

.. testcode::

>     # Both the classical and quantum registers have the same relative order as in qc

Reviewer: ```suggestion
    # Resulting quantum circuits will be the same if the quantum and classical registers have the same relative order
```

    qc1 = QuantumCircuit(qr1, cr1, qr2, cr2)

    print(qc == qc1)

.. testoutput::

    True

.. testcode::

>     # We change the order of the quantum registers

Reviewer: ```suggestion
    # Resulting quantum circuits are different if the quantum or classical registers have different relative order
```

########################
Create a quantum circuit
########################

This guide shows you how to create a quantum circuit.

There are two ways to create a :class:`~.QuantumCircuit` object:

1. Initialize :class:`~.QuantumCircuit` by directly specifying the number of qubits and classical bits you want.
2. Creating :class:`~.QuantumRegister`\ s and :class:`~.ClassicalRegister`\ s and use the registers to initialize a :class:`~.QuantumCircuit`

.. _create circuit with qubit numbers:

> Create by specifying the number of qubits and classical bits

Reviewer 2: this feels a bit weird grammatically. I'd suggest saying something like "Create a circuit by specifying..."

============================================================

You can create a :class:`~.QuantumCircuit` by only specifying the number of qubits and classical bits. For example:

.. testcode::

    from qiskit import QuantumCircuit

    # Initialize number of qubits and classical bits
    n_qubits = 3
    n_bits = 2

    # Create the circuit
    qc = QuantumCircuit(n_qubits, n_bits)
    print(qc.draw())

.. testoutput::
    :options: +NORMALIZE_WHITESPACE

    q_0: 
        
    q_1: 
        
    q_2: 
        
    c: 2/
     

You can also create a :class:`~.QuantumCircuit` with only qubits and no classical bits, by omitting the number of classical bits:

.. testcode::

    from qiskit import QuantumCircuit

    qc = QuantumCircuit(n_qubits)
    print(qc.draw())

.. testoutput::
    :options: +NORMALIZE_WHITESPACE

    q_0: 
        
    q_1: 
        
    q_2: 

> Create from quantum and classical registers

Reviewer 2: same as above, just using "create" on it's own like this feels a bit odd grammatically. I think using something like "create a circuit from..." is a bit clearer

    cr1 = ClassicalRegister(2)

    # Create a classical register with 1 classical bit
    cr2 = ClassicalRegister(1)

Create a quantum circuit
------------------------

Now that you have defined the quantum and classical registers, you can create a :class:`~.QuantumCircuit` with the registers: 

.. testcode::

    # Create the quantum circuit from the registers
    qc = QuantumCircuit(qr1, qr2, cr1, cr2)
    print(qc.draw())

.. testoutput::
    :options: +NORMALIZE_WHITESPACE

    q0_0: 
      
    q0_1: 
        
    q1_0: 
        
    q1_1: 
        
    q1_2: 
        
    c0: 2/
        
    c1: 1/
      

You can put the registers in any order, even mixing classical and quantum. However, the relative order of the :class:`~.QuantumRegister`\ s affects the order of the qubits in the final circuit. The qubits from the first :class:`~.QuantumRegister` will be the first and so on. The same applies to the :class:`~.ClassicalRegister`\ s.

.. testcode::

    # Resulting quantum circuits will be the same if the quantum and classical registers have the same relative order
    qc1 = QuantumCircuit(qr1, cr1, qr2, cr2)

    print(qc == qc1)

.. testoutput::

    True

.. testcode::

>     # Resulting quantum circuits are different if the quantum or classical registers have different relative order

Reviewer 2: nice comparison!
